pkg/operator/ceph/csi/spec.go

```diff
@@ -468,16 +479,81 @@ func startDrivers(namespace string, clientset kubernetes.Interface, ver *version
 	}
 
 	if ver.Major > KubeMinMajor || (ver.Major == KubeMinMajor && ver.Minor >= provDeploymentSuppVersion) {
-		err = createCSIDriverInfo(clientset, RBDDriverName, ownerRef)
+		if EnableRBD {
+			err = createCSIDriverInfo(clientset, RBDDriverName, ownerRef)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create CSI driver object for %q", RBDDriverName)
+			}
+		}
+		if EnableCephFS {
+			err = createCSIDriverInfo(clientset, CephFSDriverName, ownerRef)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create CSI driver object for %q", CephFSDriverName)
+			}
+		}
+	}
+	return nil
+}
+
+func stopDrivers(clientset kubernetes.Interface, namespace string, ver *version.Info) error {
+	if !EnableRBD {
+		logger.Info("CSI Ceph RBD driver disabled")
+		succeeded := deleteCSIDriverResources(clientset, namespace, ver, csiRBDPlugin, csiRBDProvisioner, "csi-rbdplugin-metrics", RBDDriverName)
+		if succeeded {
+			logger.Info("successfully removed CSI Ceph RBD driver")
+		} else {
+			logger.Error("failed to remove CSI Ceph RBD driver")
+		}
+	}
+
+	if !EnableCephFS {
+		logger.Info("CSI CephFS driver disabled")
+		succeeded := deleteCSIDriverResources(clientset, namespace, ver, csiCephFSPlugin, csiCephFSProvisioner, "csi-cephfsplugin-metrics", CephFSDriverName)
+		if succeeded {
+			logger.Info("successfully removed CSI CephFS driver")
+		} else {
+			logger.Error("failed to remove CSI CephFS driver")
+		}
+	}
+	return nil
+}
+
+func deleteCSIDriverResources(
+	clientset kubernetes.Interface, namespace string, ver *version.Info, daemonset string, deployment string, service string, driverName string) bool {
+
+	succeeded := true
+	err := k8sutil.DeleteDaemonset(clientset, namespace, daemonset)
+	if err != nil {
+		logger.Errorf("failed to delete the %q. %v", daemonset, err)
+		succeeded = false
+	}
+
+	if ver.Major == KubeMinMajor && ver.Minor < provDeploymentSuppVersion {
+		err = k8sutil.DeleteStatefulset(clientset, namespace, deployment)
 		if err != nil {
-			return errors.Wrapf(err, "failed to create CSI driver object for %q", RBDDriverName)
+			logger.Errorf("failed to delete the %q. %v", deployment, err)
+			succeeded = false
 		}
-		err = createCSIDriverInfo(clientset, CephFSDriverName, ownerRef)
+	} else if ver.Major == KubeMinMajor && ver.Minor >= provDeploymentSuppVersion {
```

Comment:
no need to check the kubernetes version again as  this will become
```go
if ver.Major == KubeMinMajor && ver.Minor < provDeploymentSuppVersion {
// delete statefulset
}else{
// delete deployment
}
```

Reply:
> no need to check the kubernetes version again as this will become
> 
> ```go
> if ver.Major == KubeMinMajor && ver.Minor < provDeploymentSuppVersion {
> // delete statefulset
> }else{
> // delete deployment
> }
> ```

done.